src/platform/packages/shared/kbn-esql-ast/src/composer/README.md

```diff
@@ -76,6 +87,433 @@ query.toRequest();
 query.toRequest().params; // [{ input : 123 }]
 ```
 
-## Reference
+## API Reference
+
+### Basic Usage Patterns
+
+The `esql` tag function supports multiple usage patterns for different scenarios:
+
+#### Tagged Template Syntax
+
+```ts
+// Static query
+const query = esql`FROM index | WHERE foo > 42 | LIMIT 10`;
+
+// Dynamic values (inlined)
+const threshold = 100;
+const query = esql`FROM index | WHERE count > ${threshold}`;
+
+// Dynamic values with parameters (recommended for security)
+const userInput = 'admin';
+const query = esql`FROM logs | WHERE user.name == ${esql.par(userInput)}`;
+```
+
+#### String Constructor
+
+```ts
+// From string with named parameters
+const query = esql('FROM index | WHERE count > ?threshold | LIMIT ?limit', {
+  threshold: 100,
+  limit: 25,
+});
+
+// Pre-parameterized function
+const builder = esql({ defaultLimit: 10 });
+const query = builder`FROM index | LIMIT ?defaultLimit`;
+```
+
+### Parameter Handling
+
+ES|QL Composer provides multiple ways to handle dynamic parameters securely:
+
+#### Explicit Parameters
+
+```ts
+// Named parameter
+const query = esql`FROM index | WHERE field > ${esql.par(value, 'threshold')}`;
+
+// Auto-generated parameter name
+const query = esql`FROM index | WHERE field > ${esql.par(value)}`;
+// Generates: WHERE field > ?p0
+```
+
+#### Shorthand Object Syntax
+
+```ts
+// Single parameter per object (enforced by TypeScript)
+const limit = 100;
+const field = 'user.name';
+const query = esql`FROM index | LIMIT ${{ limit }} | KEEP ${{ field }}`;
+// Generates: FROM index | LIMIT ?limit | KEEP ?field
```

Comment:
It should generate `.... | KEEP ??field`

The fields and functions are with double questionmarks, everything else with single

Reply:
I see. Ok so maybe we should remove it from the example as it is going to fail.

About the solutions I like all of them except for the 1st, no0t because I disagree with another syntax idea but mostly because this ` KEEP ${{ [field] }}`;` is quite confusing, at least for me